crates/orderbook-commons/src/lib.rs

```diff
@@ -172,31 +172,42 @@ pub struct FilledWith {
 }
 
 impl FilledWith {
-    /// calculates the average execution price for inverse contracts
-    ///
-    /// The average execution price follows a simple formula:
-    /// `total_order_quantity / (quantity_trade_0 / execution_price_trade_0 + quantity_trade_1 /
-    /// execution_price_trade_1 )`
     pub fn average_execution_price(&self) -> Decimal {
-        if self.matches.len() == 1 {
-            return self
-                .matches
-                .first()
-                .expect("to be exactly one")
-                .execution_price;
-        }
-
-        let sum_quantity = self
+        let amounts = self
             .matches
             .iter()
-            .fold(Decimal::ZERO, |acc, m| acc + m.quantity);
+            .map(|m| QuantityPrice {
+                quantity: m.quantity,
+                execution_price: m.execution_price,
+            })
+            .collect();
+        average_execution_price(amounts)
+    }
+}
 
-        let nominal_prices: Decimal = self.matches.iter().fold(Decimal::ZERO, |acc, m| {
-            acc + (m.quantity / m.execution_price)
-        });
+pub struct QuantityPrice {
```

Comment:
Nit: 

That's a weird struct 🙈. Maybe make two out of them and pass them in separately? 
Alternatively, pass in the match and spare the creation of the new struct?

Reply:
Unfortunately, that's not easily done. I need the relationship of the quantity to execution price. If I would split them up into two vecs I'd have to guess on the index that they relate, which is also not very nice.

Maybe improving the name of the struct helps? any suggestions?